### ISSUE TITLE
Bug: [no-unsafe-enum-comparison] false positive with enums in a union with branded types

### ISSUE THREAD

**Issue by bradzacher** [typescript-eslint/typescript-eslint#7114]
### Before You File a Bug Report Please Confirm You Have Done The Following...

- [X] I have tried restarting my IDE and the issue persists.
- [X] I have updated to the latest version of the packages.
- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22package%3A+eslint-plugin%22) and found none that matched my issue.
- [X] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### Playground Link

https://typescript-eslint.io/play/#ts=5.0.3&sourceType=module&showAST=es&code=KYOwrgtgBAYg9nKBvAsAKClAglAvFAcgEMCAadAX3XQBcBPAB2CgCEiAnPKACgGcb2ASxABzKADJkUAPrSARuyIgAJgC4oYEAGsQcAO4goFAJRQAPrAQBuammXAAxgBsOzB3BD8ocjurbsbNHRBADMeH05cKMI5AlNUDCgAeiTKdCA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6Jge1tieQEMAZonpNYAW1pkOY4nyrIOTdFETRoHaJHBgAviB1A&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA

### Repro Code

```typescript
enum Foo {
  A = 'a',
}

type Bar = (string & { __brand: unknown }) | Foo;

declare const bar: Bar;

if (bar === 'b') {
  //
}
```

### ESLint Config

```javascript
module.exports = {
  parser: "@typescript-eslint/parser",
  rules: {
    "@typescript-eslint/no-unsafe-enum-comparison": ["error"],
  },
};
```

### Expected Result

No error

### Actual Result

The rule errors on the comparison because it doesn't recognise the branded type as the base type

**Comment by Zamiell** [comment]
I'll try working on this once #7541 is approved.

**Comment by Zamiell** [comment]
I figured this one out this morning. The fix is to add this:

```ts
        const leftTypeParts1 = tsutils.unionTypeParts(left);
        const rightTypeParts1 = tsutils.unionTypeParts(right);
        const leftTypeParts2 = tsutils.intersectionTypeParts(left);
        const rightTypeParts2 = tsutils.intersectionTypeParts(right);
```

I'll submit a PR once the other one is approved.
